In [6]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [7]:
#loading the data set
df = pd.read_csv('Creditcard.csv')

In [9]:
# first 5 rows
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [11]:
# last 5 rows
df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [12]:
# data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [14]:
# checking if any missing values are there
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

<b>There is no missing values</b>

In [16]:
# distribution of class of transaction legit or fraud
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

<b>label 0 rpresents 284315 Legit or normal transaction</b>
<b> label 1 represents 492 fraud transaction </b>

<b>Data is highly unbalanced</b>

In [19]:
#seperating the data for better analysis
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [21]:
#printing the shape of the 2 classes
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [23]:
#statistical meassures
#legit.describe()
#analysis the Amount column of legit transaction
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [26]:
#analysis the Amount column of fraud transaction
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [27]:
#comparing the values of both legit and fraud transaction
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

As data set is highli imbalanced, we will create a balanced sample data set from the main data set to train the model.
Building train data set.
284315 Legit or normal transaction.
492 fraud transaction.
So we will randomly choose 492 normal transaction to match with 492 fraud transaction to prepare a balanced sample data set for a better prediction.

In [28]:
#creating 492 random legit samples
legit_sample = legit.sample(n=492)

In [31]:
# creating new data frame by concating legit_sample with fraud class
new_df = pd.concat([legit_sample, fraud], axis=0)

In [33]:
#checking the head of new data frame
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
107597  70515.0  1.177502  0.217890  0.151132  0.457431  0.022075 -0.087930   
37448   38958.0 -2.090896  1.443757  1.214177  0.148295 -1.244160 -0.190355   
43076   41359.0 -1.189499  0.754789 -0.053557 -2.894805 -1.271245 -1.512523   
44177   41828.0  1.376981 -1.209193  0.146426 -1.368474 -1.347867 -0.699607   
14976   26292.0 -2.472184  4.128632 -3.512361  0.989895 -0.210544 -1.965074   

              V7        V8        V9  ...       V21       V22       V23  \
107597 -0.092488  0.102778 -0.204125  ... -0.225313 -0.670737  0.098989   
37448   0.562739  0.727264 -0.972635  ...  0.114521 -0.064167 -0.278040   
43076  -0.298745  0.750894 -0.381743  ... -0.240205 -0.240061  0.097426   
44177  -0.756003 -0.278100 -1.914312  ... -0.184251 -0.313181 -0.115515   
14976   0.152400  0.678897  1.608722  ... -0.403628 -0.353388  0.330796   

             V24       V25       V26       V27       V28  Amount  Class  
107597 -0.348904  0.152834  0.126270 -0.012614  0.015255    9.99      0  
37448   0.516173  0.351606 -0.625616 -0.474535 -0.170199  150.00      0  
43076   0.367324 -0.264702 -0.109128  0.251213  0.136935    4.00      0  
44177  -0.116427  0.440463 -0.187123  0.015922  0.033252  105.74      0  
14976   0.109839  0.089482 -0.398126  1.085204  0.381063    0.89      0  

[5 rows x 31 columns]

In [34]:
#checking the tail of the new data frame
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [38]:
#counting of both legit and fraud transaction
new_df['Class'].value_counts()

1    492
0    492
Name: Class, dtype: int64

In [39]:
#comparing the values of both legit and fraud transaction
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      97492.445122  0.041128  0.033941 -0.103172 -0.043256 -0.025138   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.022715 -0.016370 -0.002738  0.019609  ... -0.028072  0.011112   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.065003  0.020129  0.008153 -0.040365  0.008216  0.039468  0.001386   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       83.837215  
1      122.211321  

[2 rows x 30 columns]

Spliting the data into Features & Targets

In [40]:
#dropping the predicting features, The class column
X = new_df.drop(columns='Class', axis=1)
Y = new_df['Class']

In [41]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
107597   70515.0  1.177502  0.217890  0.151132  0.457431  0.022075 -0.087930   
37448    38958.0 -2.090896  1.443757  1.214177  0.148295 -1.244160 -0.190355   
43076    41359.0 -1.189499  0.754789 -0.053557 -2.894805 -1.271245 -1.512523   
44177    41828.0  1.376981 -1.209193  0.146426 -1.368474 -1.347867 -0.699607   
14976    26292.0 -2.472184  4.128632 -3.512361  0.989895 -0.210544 -1.965074   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [43]:
print(Y)

107597    0
37448     0
43076     0
44177     0
14976     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


<b>Splitting the Data into Training and Testing Data</b>

In [44]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [46]:
#checking the spliting
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


<b>Modele Traing with Logistic Regression</b>

In [47]:
#creating hte model
model = LogisticRegression()

In [48]:
# training the model with training data
model.fit(X_train, Y_train)

LogisticRegression()

<b>Model Evalution</b>

In [54]:
# finding the accuracy score of training data
X_train_predict = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_predict, Y_train)
#printing the accuracy of training data
print('Accuracy of Training data = ',training_data_accuracy)

Accuracy of Training data =  0.9453621346886912


Prediction is good of Traing data, more than 75%

In [55]:
# finding the accuracy score of test data
X_test_predict = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_predict, Y_test)
#printing the accuracy of test data
print('Accuracy of Test data = ',testing_data_accuracy)

Accuracy of Test data =  0.9137055837563451


Prediction is good of Testing data, more than 75%